Mathlib/Analysis/Convex/Deriv.lean

```diff
@@ -366,3 +371,319 @@ theorem strictConcaveOn_univ_of_deriv2_neg {f : ℝ → ℝ} (hf : Continuous f)
     (hf'' : ∀ x, deriv^[2] f x < 0) : StrictConcaveOn ℝ univ f :=
   strictConcaveOn_of_deriv2_neg' convex_univ hf.continuousOn fun x _ => hf'' x
 #align strict_concave_on_univ_of_deriv2_neg strictConcaveOn_univ_of_deriv2_neg
+
+/-!
+## Convexity of `f` implies monotonicity of `f'`
+
+In this section we prove inequalities relating derivatives of convex functions to slopes of secant
+lines, and deduce that if `f` is convex then its derivative is monotone (and similarly for strict
+convexity / strict monotonicity).
+-/
+
+/-- A non-singleton convex set in `ℝ` has the `UniqueDiffWithinAt` property, i.e. derivatives
+within `S` are unique when they exist. -/
+lemma Convex.uniqueDiffWithinAt {S : Set ℝ} (hS : Convex ℝ S) {x y : ℝ}
+    (hx : x ∈ S) (hy : y ∈ S) (hxy : x ≠ y) :
+    UniqueDiffWithinAt ℝ S x := by
+  -- Show that `𝓝[S] x` is always bounded below by either `𝓝[<] x` or `𝓝[>] x`.
+  rcases lt_or_lt_iff_ne.mpr hxy with hxy | hxy
+  · apply (uniqueDiffWithinAt_Ioi _).mono_nhds
+    rw [nhdsWithin_le_iff, mem_nhdsWithin_Ioi_iff_exists_Ioc_subset]
+    refine ⟨y, hxy, Ioc_subset_Icc_self.trans <| hS.ordConnected.out hx hy⟩
+  · apply (uniqueDiffWithinAt_Iio _).mono_nhds
+    rw [nhdsWithin_le_iff, mem_nhdsWithin_Iio_iff_exists_Ico_subset]
+    refine ⟨y, hxy, Ico_subset_Icc_self.trans <| hS.ordConnected.out hy hx⟩
```

Comment:
I'm afraid that the `Nontrivial` hypothesis is definitely more annoying later. Currently this is used as
`hfc.1.uniqueDiffWithinAt hy hx hxy.ne'` (inside an argument of a simp lemma). With this suggested reformulation, it would become
````
(uniqueDiffOn_ordConnected_of_nontrivial hfc.1.ordConnected 
      (Set.nontrivial_of_exists_lt ⟨_, hx, _, hy, hxy⟩)) y hy)
````
which is far more annoying.

(Note that I'm deliberately re-using `x` here: it's both the point at which we want the derivative to be unique, and one of the elements of `S` we use to establish nontriviality. This is order to minimise the number of arguments to be supplied at the point of use, and you can't do this if you go via `UniqueDiffOn`.)